File: 025-二叉树中和为某一值的路径/025-FindPath.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
import copy
class Solution:
    # 返回二维列表，内部每个列表表示找到的路径
    def __init__(self):
        self.ans=[]
    def FindPath(self, root, expectNumber):
        # write code here
        if root is None:
            return []
        self.find(root,expectNumber,0,[])
        return self.ans
    def find(self,root,expectNumber,cur,path):
        cur+=root.val
        path.append(root.val)
        if cur==expectNumber and root.left is None and root.right is None:
            self.ans.append(path)
        if root.left is not None:
            self.find(root.left,expectNumber,cur,copy.copy(path))
        if root.right is not None:
            self.find(root.right,expectNumber,cur,copy.copy(path))

File: 025-二叉树中和为某一值的路径/025-FindPath.cpp
/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution {
    vector<vector<int>> ans;
public:
    vector<vector<int> > FindPath(TreeNode* root,int expectNumber) {
        if(root == nullptr)
        {
            return ans;
        }
        vector<int> path;
        find(root, expectNumber, path, 0);

        return ans;
    }
    void find(TreeNode* root,int expectNumber,vector<int> path,int cur){
        cur+=root->val;
        path.push_back(root->val);
        if (cur==expectNumber && root->left==nullptr && root->right==nullptr){
            ans.push_back(path);
            return ;
        }
        if(root->left != nullptr)
        {
            find(root->left, expectNumber, path, cur);
        }
        if(root->right != nullptr)
        {
            find(root->right, expectNumber, path, cur);
        }
    }
};
